### ISSUE TITLE
v12 bug with zip files

### ISSUE THREAD

**Issue by rcohenma** [hashicorp/terraform#21868]
### Terraform Version

```
2019/06/24 17:04:23 [INFO] Terraform version: 0.12.2  
2019/06/24 17:04:23 [INFO] Go runtime version: go1.12.4
2019/06/24 17:04:23 [INFO] CLI args: []string{"/Users/rcohenma/terraform12", "-v"}
2019/06/24 17:04:23 [DEBUG] Attempting to open CLI config file: /Users/rcohenma/.terraformrc
2019/06/24 17:04:23 Loading CLI configuration from /Users/rcohenma/.terraformrc
2019/06/24 17:04:23 [INFO] CLI command args: []string{"version", "-v"}
Terraform v0.12.2
2019/06/24 17:04:23 [DEBUG] checking for provider in "."
2019/06/24 17:04:23 [DEBUG] checking for provider in "/Users/rcohenma"
2019/06/24 17:04:23 [DEBUG] checking for provider in ".terraform/plugins/darwin_amd64"
2019/06/24 17:04:23 [DEBUG] found provider "terraform-provider-local_v1.2.2_x4"
2019/06/24 17:04:23 [DEBUG] found provider "terraform-provider-random_v2.1.2_x4"
2019/06/24 17:04:23 [DEBUG] checking for provider in "/Users/rcohenma/.terraform.d/plugins"
2019/06/24 17:04:23 [DEBUG] checking for provider in "/Users/rcohenma/.terraform.d/plugins/darwin_amd64"
2019/06/24 17:04:23 [DEBUG] found plugin override in .terraformrc: "oci", "/Users/rcohenma/GoglandProjects/bin/terraform-provider-oci"
2019/06/24 17:04:23 [DEBUG] found valid plugin: "random", "2.1.2", "/Users/rcohenma/GoglandProjects/src/github.com/terraform-providers/terraform-provider-oci/examples/database/adb/adw/.terraform/plugins/darwin_amd64/terraform-provider-random_v2.1.2_x4"
2019/06/24 17:04:23 [DEBUG] found valid plugin: "oci", "0.0.0", "/Users/rcohenma/GoglandProjects/bin/terraform-provider-oci"
2019/06/24 17:04:23 [DEBUG] found valid plugin: "local", "1.2.2", "/Users/rcohenma/GoglandProjects/src/github.com/terraform-providers/terraform-provider-oci/examples/database/adb/adw/.terraform/plugins/darwin_amd64/terraform-provider-local_v1.2.2_x4"
+ provider.local v1.2.2
+ provider.oci (unversioned)
+ provider.random v2.1.2
```

### Expected Behavior
v0.12.x should be able to store zip files in its statefile without corrupting them as v0.11.x does

### Actual Behavior
When the content of a zip file is written to a file it is corrupted and customers cannot unzip it

### Steps to Reproduce

In this example https://github.com/terraform-providers/terraform-provider-oci/blob/master/examples/database/adb/autonomous_data_warehouse_wallet.tf#L8
"data.oci_database_autonomous_database_wallet.autonomous_data_warehouse_wallet" returns a zip file which is stored in a field called "content".
The example writes this content into a file `autonomous_data_warehouse_wallet.zip`
running `unzip  autonomous_data_warehouse_wallet.zip` after the apply results in an error when using v0.12.x

### Additional Context

I am a developer on the OCI provider. 

v12 has a bug when storing zip files in the statefile

We have a service operation that returns a zip file and we have a data_source for it.

The file gets corrupted when stored on the statefile and our customer are seeing errors when they  attempt to unzip it.

With v11 it works perfectly so this is an issue with v12 specifically

Could you fix this bug?

In the mean time could you advise on a potential workaround?

Thanks

**Comment by alexng-canuck** [comment]
It appears the new shims in v0.12 plugin SDKs will do some NFC normalization of strings written by a provider before persisting to state. I wonder if this is what's causing the regression in behavior between v0.11 and v0.12.

**Comment by paultyng** [comment]
0.11 didn't intend to support binary data, and there was corruption using some of the code paths. It worked in others, and this is due mostly to Go representing strings under the hood as slices of bytes. This case happened to work, but was not intended to do so. In 0.12 we have tightened up the encoding use cases, but we have also introduced more methods that work on base64 encoded binary data. 

OCI should output this content as a base64 encoded string attribute to ensure compatibility with all of Terraform core.

For the use case with the `local` provider, I have added `content_base64` support to the local provider to allow writing base64 encoded binary data to a file directly: https://github.com/terraform-providers/terraform-provider-local/pull/29
